diff -Naur procps-3.2.8.orig/sysctl.c procps-3.2.8/sysctl.c
--- procps-3.2.8.orig/sysctl.c	2011-08-02 19:31:54.000000000 +0200
+++ procps-3.2.8/sysctl.c	2011-08-02 19:45:09.203257948 +0200
@@ -190,6 +190,19 @@
    }
 
    if (stat(tmpname, &st)==0) {
+
+	/* Bug 697935 - [RFE] Restore support for partial sysctl keys */
+	if (S_ISDIR(st.st_mode)) {
+	    size_t len;
+	    len = strlen(tmpname);
+	    tmpname[len] = '/';
+	    tmpname[len+1] = '\0';
+	    rc = DisplayAll(tmpname);
+	    free(tmpname);
+	    free(outname);
+	    return rc;
+	}
+
 	if (st.st_mode & (S_IRUSR|S_IROTH|S_IRGRP))	   
 	   	fp = fopen(tmpname, "r");
 	else
