diff -up procps-3.2.8/vmstat.c.long procps-3.2.8/vmstat.c
--- procps-3.2.8/vmstat.c.long	2011-03-31 11:11:38.000000000 +0200
+++ procps-3.2.8/vmstat.c	2011-05-18 11:54:51.395334800 +0200
@@ -183,7 +183,7 @@ static void new_header(void){
 
 ////////////////////////////////////////////////////////////////////////////
 
-static unsigned long unitConvert(unsigned int size){
+static unsigned long unitConvert(unsigned long size){
  float cvSize;
  cvSize=(float)size/dataUnit*((statMode==SLABSTAT)?1:1024);
  return ((unsigned long) cvSize);
