diff -Naur procps-3.2.8.orig/top.c procps-3.2.8/top.c
--- procps-3.2.8.orig/top.c	2014-06-24 16:39:26.000000000 +0200
+++ procps-3.2.8/top.c	2014-06-24 17:02:14.000000000 +0200
@@ -1884,8 +1884,15 @@
                else if (*args) cp = *args++;
                else std_err("-d requires argument");
                   /* a negative delay will be dealt with shortly... */
-               if (sscanf(cp, "%f", &tmp_delay) != 1)
-                  std_err(fmtmk("bad delay '%s'", cp));
+               char *str;
+               tmp_delay = strtof(cp, &str);
+               if (*str) {
+                  /* fallback - try to swap the floating point separators */
+                  if (*str == '.') (*str)=',';
+                  else if (*str == ',') (*str)='.';
+                  tmp_delay = strtof(cp, &str);
+                  if (*str) std_err(fmtmk("bad delay '%s'", cp));
+               }
                break;
             case 'H':
                TOGw(Curwin, Show_THREADS);
diff -Naur procps-3.2.8.orig/watch.c procps-3.2.8/watch.c
--- procps-3.2.8.orig/watch.c	2014-06-24 16:39:25.000000000 +0200
+++ procps-3.2.8/watch.c	2014-06-24 16:51:00.000000000 +0200
@@ -193,8 +193,14 @@
 			{
 				char *str;
 				interval = strtod(optarg, &str);
-				if (!*optarg || *str)
-					do_usage();
+				if (!*optarg) do_usage();
+				if (*str) {
+					/* fallback - try to swap the floating point separators */
+					if (*str == '.') (*str)=',';
+					else if (*str == ',') (*str)='.';
+					interval = strtod(optarg, &str);
+					if (*str) do_usage();
+				}
 				if(interval < 0.1)
 					interval = 0.1;
 				if(interval > ~0u/1000000)
