diff -Naur procps-3.2.8.orig/proc/readproc.c procps-3.2.8/proc/readproc.c
--- procps-3.2.8.orig/proc/readproc.c	2015-03-02 14:16:46.000000000 +0100
+++ procps-3.2.8/proc/readproc.c	2015-03-02 15:16:16.126003524 +0100
@@ -352,7 +352,9 @@
     P->sched = -1;
     P->nlwp = 0;
 
-    S = strchr(S, '(') + 1;
+    S = strchr(S, '(');
+    if(unlikely(!S)) goto out;
+    S++;
     tmp = strrchr(S, ')');
     num = tmp - S;
     if(unlikely(num >= sizeof P->cmd)) num = sizeof P->cmd - 1;
@@ -399,6 +401,7 @@
       P->nlwp = 1;
     }
 
+out:
 LEAVE(0x160);
 }
 
