diff --git a/ps/module.mk b/ps/module.mk
index 2902a3a..998ae77 100755
--- a/ps/module.mk
+++ b/ps/module.mk
@@ -16,11 +16,13 @@ PSNAMES := $(addprefix ps/,$(PS_C))
 PSOBJ   := $(addsuffix .o,$(PSNAMES))
 PSSRC   := $(addsuffix .c,$(PSNAMES))
 
+SELINUX_LDFLAGS = -lselinux
+
 PS_X := COPYING HACKING TRANSLATION common.h module.mk it p ps.1 regression
 TARFILES += $(PSSRC) $(addprefix ps/,$(PS_X))
 
 ps/ps: $(PSOBJ) $(LIBPROC)
-	$(CC) $(ALL_CFLAGS) $(ALL_LDFLAGS) -o $@ $^ $(ldl)
+	$(CC) $(ALL_CFLAGS) $(ALL_LDFLAGS) $(SELINUX_LDFLAGS) -o $@ $^ $(ldl)
 
 # This just adds the stacktrace code
 ps/debug: $(PSOBJ) stacktrace.o $(LIBPROC)
diff --git a/ps/output.c b/ps/output.c
index 9ad6007..9454128 100644
--- a/ps/output.c
+++ b/ps/output.c
@@ -54,6 +54,7 @@
 #include <time.h>
 #include <unistd.h>
 #include <dlfcn.h>
+#include <selinux/selinux.h>
 
 #include "../proc/readproc.h"
 #include "../proc/sysinfo.h"
@@ -1107,6 +1108,7 @@ static int pr_sgi_p(char *restrict const outbuf, const proc_t *restrict const pp
 }
 
 
+#if 0
 /****************** FLASK & seLinux security stuff **********************/
 // move the bulk of this to libproc sometime
 
@@ -1138,28 +1140,15 @@ fail:
   return 1;
 }
 
-#if 0
+#else
 // This needs more study, considering:
-// 1. the static linking option (maybe disable this in that case)
-// 2. the -z and -Z option issue
-// 3. width of output
+// 1. the -z and -Z option issue
+// 2. width of output
 static int pr_context(char *restrict const outbuf, const proc_t *restrict const pp){
-  static int (*ps_getpidcon)(pid_t pid, char **context) = 0;
-  static int tried_load = 0;
   size_t len;
   char *context;
 
-  if(!ps_getpidcon && !tried_load){
-    void *handle = dlopen("libselinux.so.1", RTLD_NOW);
-    if(handle){
-      dlerror();
-      ps_getpidcon = dlsym(handle, "getpidcon");
-      if(dlerror())
-        ps_getpidcon = 0;
-    }
-    tried_load++;
-  }
-  if(ps_getpidcon && !ps_getpidcon(pp->tgid, &context)){
+  if(!getpidcon(pp->tgid, &context)){
     size_t max_len = OUTBUF_SIZE-1;
     len = strlen(context);
     if(len > max_len) len = max_len;
