diff -Naur procps-3.2.8.orig/top.c procps-3.2.8/top.c
--- procps-3.2.8.orig/top.c	2011-12-08 16:09:58.000000000 +0100
+++ procps-3.2.8/top.c	2011-12-08 16:11:57.532648486 +0100
@@ -1861,7 +1861,7 @@
       .  bunched args are actually handled properly and none are ignored
       .  we tolerate NO whitespace and NO switches -- maybe too tolerant? */
    static const char usage[] =
-      " -hv | -bcisSHM -d delay -n iterations [-u user | -U user] -p pid [,pid ...]";
+      " -hv | -abcHimMsS -d delay -n iterations [-u user | -U user] -p pid [,pid ...]";
    float tmp_delay = MAXFLOAT;
    char *p;
 
@@ -1963,6 +1963,9 @@
             case 'M':
                Mem_units_detect = 1;
                break;
+            case 'a':
+               Curwin->rc.sortindx = P_MEM;
+               break;
             default :
                std_err(fmtmk("unknown argument '%c'\nusage:\t%s%s"
                   , *cp, Myname, usage));
