diff -Naur procps-3.2.8.orig/pgrep.c procps-3.2.8/pgrep.c
--- procps-3.2.8.orig/pgrep.c	2007-05-28 02:11:07.000000000 +0200
+++ procps-3.2.8/pgrep.c	2014-06-17 19:49:28.919457769 +0200
@@ -697,11 +697,11 @@
         if (argc - optind == 1)
 		opt_pattern = argv[optind];
 	else if (argc - optind > 1)
-		usage (0);
+		usage ('?');
 	else if (criteria_count == 0) {
 		fprintf (stderr, "%s: No matching criteria specified\n",
 			 progname);
-		usage (0);
+		usage ('?');
 	}
 }
 
