diff -Naur procps-3.2.8.orig/tload.c procps-3.2.8/tload.c
--- procps-3.2.8.orig/tload.c	2002-10-06 23:21:38.000000000 +0200
+++ procps-3.2.8/tload.c	2011-09-15 18:30:05.634034335 +0200
@@ -43,6 +43,7 @@
 static void setsize(int i)
 {
     struct winsize win;
+    char *newscreen;
 
     signal(SIGWINCH, setsize);
     if (ioctl(fd, TIOCGWINSZ, &win) != -1) {
@@ -52,15 +53,23 @@
 	    nrows = win.ws_row;
     }
     scr_size = nrows * ncols;
-    if (screen == NULL)
-    	screen = (char *) malloc(scr_size);
-    else
-    	screen = (char *) realloc(screen, scr_size);
 
     if (screen == NULL) {
-	perror("");
-	exit(1);
+	screen = (char *) malloc(scr_size);
+	if (screen == NULL) {
+	    perror("");
+	    exit(1);
+	}
+    } else {
+	newscreen = (char *) realloc(screen, scr_size);
+	if (newscreen == NULL) {
+	    free(screen);
+	    perror("");
+	    exit(1);
+	}
+	screen = newscreen;
     }
+
     memset(screen, ' ', scr_size-1);
     *(screen + scr_size - 2) = '\0';
     if (i)
