diff -Naur procps-3.2.8.orig/sysctl.c procps-3.2.8/sysctl.c
--- procps-3.2.8.orig/sysctl.c	2014-07-16 19:03:57.000000000 +0200
+++ procps-3.2.8/sysctl.c	2014-07-16 19:04:20.814330554 +0200
@@ -542,10 +542,14 @@
          case 'f':  // the NetBSD way
          case 'p':
               argv++;
-              if (argv && *argv && **argv) {
+              ReturnCode = 0;
+              while (argv && *argv && **argv) {
                  preloadfile = *argv;
+                 ReturnCode |= Preload(preloadfile);
+                 argv++;
               }
-              return Preload(preloadfile);
+              if (preloadfile == DEFAULT_PRELOAD) ReturnCode = Preload(preloadfile);
+              return ReturnCode;
 	 case 'q':
 	      Quiet = true;
 	   break;
diff -Naur procps-3.2.8.orig/sysctl.8 procps-3.2.8/sysctl.8
--- procps-3.2.8.orig/sysctl.8	2014-07-16 19:03:57.000000000 +0200
+++ procps-3.2.8/sysctl.8	2014-07-16 19:19:30.094325334 +0200
@@ -14,7 +14,7 @@
 .br
 .B "sysctl [-n] [-e] [-q] -w variable=value ..."
 .br
-.B "sysctl [-n] [-e] [-q] -p <filename>"
+.B "sysctl [-n] [-e] [-q] -p [filename]..."
 .br
 .B "sysctl [-n] [-e] -a"
 .br
@@ -56,7 +56,7 @@
 Use this option when you want to change a sysctl setting.
 .TP
 .B "-p"
-Load in sysctl settings from the file specified or /etc/sysctl.conf if none given.
+Load in sysctl settings from the file(s) specified or /etc/sysctl.conf if none given.
 Specifying \- as filename means reading data from standard input.
 .TP
 .B "-a"
