diff -Naur procps-3.2.8.orig/vmstat.c procps-3.2.8/vmstat.c
--- procps-3.2.8.orig/vmstat.c	2012-09-25 16:46:45.000000000 +0200
+++ procps-3.2.8/vmstat.c	2012-11-13 17:34:50.085432417 +0100
@@ -236,6 +236,10 @@
   dstl= *cpu_zzz;
   Div= duse+dsys+didl+diow+dstl;
   divo2= Div/2UL;
+
+  /* SIGFPE prevention - platform specific issue */
+  if (!Div) Div = 1;
+
   printf(f,
 	 running, blocked,
 	 unitConvert(kb_swap_used), unitConvert(kb_main_free),
@@ -293,6 +297,10 @@
 
     Div= duse+dsys+didl+diow+dstl;
     divo2= Div/2UL;
+
+    /* SIGFPE prevention - platform specific issue */
+    if (!Div) Div = 1;
+
     printf(f,
            running, blocked,
 	   unitConvert(kb_swap_used),unitConvert(kb_main_free),
